thrust/thrust/iterator/counting_iterator.h

```diff
@@ -46,91 +46,42 @@
-//      sizeof(Integer) < sizeof(std::ptrdiff_t),
-//      identity_<std::ptrdiff_t>,
-//      identity_<intmax_t>
-//    >
-//  >
-{
-private:
-
-public:
-  using type =
-    typename eval_if<::cuda::std::numeric_limits<Integer>::is_signed
-                       && (!::cuda::std::numeric_limits<Integer>::is_bounded
-                           || (int(::cuda::std::numeric_limits<Integer>::digits) + 1 >= num_digits<intmax_t>::value)),
-                     identity_<Integer>,
-                     eval_if<int(::cuda::std::numeric_limits<Integer>::digits) + 1 < num_digits<int>::value,
-                             identity_<int>,
-                             eval_if<int(::cuda::std::numeric_limits<Integer>::digits) + 1 < num_digits<long>::value,
-                                     identity_<long>,
-                                     identity_<intmax_t>>>>::type;
-}; // end integer_difference
-
-template <typename Number>
-struct numeric_difference
-    : eval_if<::cuda::std::is_integral<Number>::value, integer_difference<Number>, identity_<Number>>
-{}; // end numeric_difference
-
 template <typename Number>
-_CCCL_HOST_DEVICE typename numeric_difference<Number>::type numeric_distance(Number x, Number y)
-{
-  using difference_type = typename numeric_difference<Number>::type;
-  return difference_type(y) - difference_type(x);
-} // end numeric_distance
+using counting_iterator_difference_type = ::cuda::std::_If<
+  ::cuda::std::is_integral_v<Number>,
+  // the difference between two int values can be larger than what an int can represent
+  ::cuda::std::_If<sizeof(Number) < sizeof(int), int, ::cuda::std::ptrdiff_t>,
+  // floating points use ptrdiff_t
+  ::cuda::std::_If<::cuda::std::is_floating_point_v<Number>,
+                   ::cuda::std::ptrdiff_t,
+                   // any other type, if it can represent the difference, can be used as difference type,
+                   // otherwise also ptrdiff_t
+                   ::cuda::std::_If<::cuda::std::numeric_limits<Number>::is_signed
+                                      && (!::cuda::std::numeric_limits<Number>::is_bounded
+                                          || ::cuda::std::numeric_limits<Number>::digits
+                                               > ::cuda::std::numeric_limits<::cuda::std::ptrdiff_t>::digits),
+                                    Number,
+                                    ::cuda::std::ptrdiff_t>>>;
 
 template <typename Incrementable, typename System, typename Traversal, typename Difference>
 struct make_counting_iterator_base
 {
   using system =
     typename eval_if<::cuda::std::is_same<System, use_default>::value, identity_<any_system_tag>, identity_<System>>::type;
 
-  using traversal =
-    typename ia_dflt_help<Traversal,
-                          eval_if<is_numeric<Incrementable>::value,
-                                  identity_<random_access_traversal_tag>,
-                                  iterator_traversal<Incrementable>>>::type;
```

Comment:
It seems a bit that `counting_iterator` used to work on iterators as incrementable types. We had no tests for this, so I wonder whether we want to keep supporting that.

Reply:
That turned out to be useful, we should keep it.